odoo/fields.py

```diff
@@ -2863,23 +2864,27 @@ def _related_domain(self):
     _description_context = property(attrgetter('context'))
 
     def _description_domain(self, env):
-        if self.check_company and not self.domain:
+        company_domain = None
+        if self.check_company:
             if self.company_dependent:
                 if self.comodel_name == "res.users":
                     # user needs access to current company (self.env.company)
-                    return "[('company_ids', 'in', allowed_company_ids[0])]"
+                    company_domain = "[('company_ids', 'in', allowed_company_ids[0])]"
                 else:
-                    return "[('company_id', 'in', [allowed_company_ids[0], False])]"
+                    company_domain = "[('company_id', 'in', [allowed_company_ids[0], False])]"
             else:
                 # when using check_company=True on a field on 'res.company', the
                 # company_id comes from the id of the current record
                 cid = "id" if self.model_name == "res.company" else "company_id"
                 if self.comodel_name == "res.users":
                     # User allowed company ids = user.company_ids
-                    return f"['|', (not {cid}, '=', True), ('company_ids', 'in', [{cid}])]"
+                    company_domain = f"['|', (not {cid}, '=', True), ('company_ids', 'in', [{cid}])]"
                 else:
-                    return f"[('company_id', 'in', [{cid}, False])]"
-        return self.domain(env[self.model_name]) if callable(self.domain) else self.domain
+                    company_domain = env[self.comodel_name].check_company_domain(companies=unquote(f"{cid} or current_company_id"))
```

Comment:
If the company field in the form is not set, the domain should not constrain the value.
And dropping `current_company_id` here does not look correct to me.
The company field in the form should be the reference company.
The "current company" is only useful for getting a default value.